Enhancements for Tutorial Notebooks and Comprehensive Support for AnnData

katosh opened the pull request:
This Pull Request (PR) consolidates all the proposed modifications aimed at improving the tutorial notebooks we are preparing for the Mellon publication. The intention is to foster constructive discussions and considerations about these changes before submitting the PR to the main repository, [dpeerlab/Palantir](https://github.com/dpeerlab/Palantir).

A major element of this PR is the introduction of enhanced support for AnnData objects. This upgrade standardizes and automates our established best practices when working with AnnData and Palantir, and ensures that these improvements don't compromise the existing functionalities.

The principal commit incorporating these extensive changes is 3607b44b5d0583be843b833fba85267c934a1c00.

katosh committed:
remove and ignore .DS_Store

katosh committed:
code formatting and linting

katosh committed:
implement `palantir.utils.early_cell`

katosh committed:
implement `palantir.utils.find_terminal_states`

This is for consideration as there already is a function
`_terminal_states_from_markov_chain`. The advantage of the
new funcion beeing that specific cell types can be defined
for which a terminal state ought to be found.

katosh committed:
checks for early_cell

katosh committed:
ignore .h5ad

katosh committed:
Add AnnData support and configuration for key inputs and outputs

This commit introduces comprehensive compatibility with the AnnData object.
The following major functions are updated to accept AnnData as an input,
perform computations, and write results back to the AnnData object:

- core.run_palantir
- utils.run_pca
- utils.run_diffusion_maps
- run_magic_imputation
- utils.determine_multiscale_space
- utils.run_tsne
- utils.determine_cell_clusters

Key enhancements and modifications include:

- All key input and output references within the AnnData structure are now customizable, providing flexibility for users to align with their existing data structures.
- Detailed docstrings have been updated to provide clear and comprehensive explanations for all new functionalities.
- Backwards compatibility has been ensured, and the modified functions have been tested with the [tutorial notebook](https://github.com/dpeerlab/Palantir/blob/master/notebooks/Palantir_sample_notebook.ipynb) to ensure the integrity and reliability of the updates.

These changes aim to increase user-friendliness and perpare the implementation of furher features.

ManuSetty commented:
Added some comments in the commit linked above - happy to discuss further!

katosh committed:
bugfixes and finilizations of sc.AnnData support

katosh committed:
implement select_branch_cells

katosh committed:
support sc.AnnData in `cluster_gene_trends`

katosh committed:
gitignore pip install build data

katosh committed:
replace PhenoGraph with Leiden

PhenoGraph fails to cluster with current package versions:
 - Python==3.10.10
 - scipy==1.10.1
 - numpy==1.24.3
 - pandas==1.5.3
 - PhenoGraph==git+https://github.com/JonathanShor/PhenoGraph@f63b02c7f5e6f2300156d7e4201818fa238bf0a4

The error is `TypeError: Expected list, got numpy.ndarray` happending in
`scipy/sparse/_lil.py:473` when calling
`_csparsetools.lol_flatten_to_array(self.rows, indices)`.

katosh commented:
Rplacing PhenoGraph with leiden in 4f935642cbe19d618b3ea599041fb30e0085de63 changes the results of the tutorial notebook since leiden finds 10 clusters with its default resolution:

![image](https://github.com/settylab/Palantir/assets/4857068/b317c8d2-663d-44c5-8498-18db346f9ee4)

katosh committed:
Implement Manu's comments

 - bugfix: use pca_key
 - parallel MAGIC on optionally sparse input
 - remove `palantir.utils.run_tsne`

katosh commented:
`palantir.utils.run_magic_imputation` now takes any input:
![image](https://github.com/settylab/Palantir/assets/4857068/56a51af6-f78d-4a30-b696-30b016f318e0)

ManuSetty commented:
> Rplacing PhenoGraph with leiden in [4f93564](https://github.com/settylab/Palantir/commit/4f935642cbe19d618b3ea599041fb30e0085de63) changes the results of the tutorial notebook since leiden finds 10 clusters with its default resolution:
> 
> ![image](https://user-images.githubusercontent.com/4857068/246950331-b317c8d2-663d-44c5-8498-18db346f9ee4.png)

These clusters do look quite good to me.

ManuSetty commented on src/palantir/presults.py:
```diff
@@ -195,45 +253,79 @@ def _gam_fit_predict_rpy2(x, y, weights=None, pred_x=None):
-    :param k: K for nearest neighbor construction
-    :param n_jobs: Number of jobs for parallel processing
-    :return: Clustering of gene trends
+def cluster_gene_trends(
```
Can we please add an optional `var` key if Anndata is specified to indicate which genes need to be clustered? I dont think we would want to cluster all genes at any point.

katosh replied:
With bb579cc72e3c6594530624ea3ca5bed9c07725bb the function now takes a gene list in the `genes` argument.

ManuSetty commented on src/palantir/utils.py:
```diff
@@ -113,81 +196,417 @@ def run_diffusion_maps(data_df, n_components=10, knn=30, alpha=0, seed=None):
 
-def run_magic_imputation(data, dm_res, n_steps=3):
-    """Run MAGIC imputation
+def _dot_herlper_func(x, y):
```
Small typo here. `herlper` -> `helper`

katosh committed:
implement mellon gene trend computation

katosh commented:
The last commit 68af21fb8c777a4708c3a8b3320532e1951e7b9e depends on a version of mellon that has not been released yet. Please install it with
```
pip install --no-deps 'git+https://github.com/settylab/Palantir@mellon'
pip install 'git+https://github.com/settylab/Mellon'
```

katosh committed:
herlper typo

katosh committed:
compare gene trends between branches

This commit enables tthe comparison while using the new implementation
that uses mellon for gene trends and saves all results in the procided
Anntata.

To finilize the new workflow of passing on AnnData I udpated all
functuions used in the tutorial and updated the tutorial itself
with the new workflow.

Since we decided to remove unused functions such as run_tsne I also
removed the following functions that are covered by scanpy:
 - palantir.utils.determine_cell_clusters
 - palantir.plot.plot_cell_clusters

katosh commented:
- Gene trends can now be compared again.
 - I also finished implementing the AnnData centered workflow throughout all functions used in the [tutorial notebook](https://github.com/settylab/Palantir/blob/mellon/notebooks/Palantir_sample_notebook.ipynb).
 - I added type hints and extended the documentation of all touched functions.
 - While almost everything is backwards compatible with the previous workflow of passing dataframes and dictionaries, I updated the tutorial to reflect the new AnnData centered workflow for ease of use.
 - Since we removed `run_tsne` (since scanpy covers this) I also removed `palantir.utils.determine_cell_clusters` and `palantir.plot.plot_cell_clusters` for the same reason

For testing install with
```
pip install --no-deps 'git+https://github.com/settylab/Palantir@mellon'
pip install 'git+https://github.com/settylab/Mellon'
```

@ManuSetty This is all from my side for now. Do you have further suggestions? When we release Mellon v1.3.0 (a new dependency) then we can make the PR to the official repo if you like.

katosh committed:
lineages option for `compute_gene_trends`

katosh committed:
small updates and cleanup of tutorial notebook

katosh committed:
bump version 1.1 -> 1.2

katosh committed:
make palantir trsults fully reproducible

katosh committed:
remove np.random.seed from tutorial since it is not needed anymore

ManuSetty commented:
I just noticed in cell 2 of the notebooks that the anndata load says `pbmc_10x_rna_processed.h5ad` - this should be `marrow_counts...`

katosh committed:
use correct dataset name in tutorial

katosh commented:
You are right! Fixed it with 99f138d9c87b7036b2a306d50d7802a3802327b9

katosh committed:
update and v1.2.0 release notes in readme

katosh commented:
Mellon v1.3.0 is now available in pypi: https://pypi.org/project/mellon/ so the new dependency is fulfillable.

The last thing before merge should be the implementation of local gene variability.
